k_pipe: add-ons to the k_pipe API rewrite

npitre opened the pull request:
This implements various cleanups and optimizations I saw the potential for
when I reviewed PR #83283. The original author did quite a lot of good work
already and I didn't want to block the merging of his own PR with these any
longer. Furthermore, it becomes easier to convey those things directly in
the code at some point.

Then it was pretty easy to add back the ability to do direct writer-to-reader
transfers and no-ring-buffer operations as the deprecated code did.

npitre committed:
kernel/pipe: code flow cleanup

Simplify the logic, avoid repeated conditionals, avoid superfluous
scheduler calls, make the code more efficient and easier to read.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
kernel/pipe: remove extra hardware timer query on thread wakeup

Dispense with the call to sys_timepoint_expired() by leveraging
swap_retval to distinguish between notifications and timeouts when
z_pend_curr() returns.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
kernel/pipe: implement direct-to-pending-readers data copy

If there are pending readers, it is best to perform a single data copy
directly into their final destination buffer rather than doing one copy
into the ring buffer just to immediately copy the same data out of it.

Incidentally, this allows for supporting pipes with no ring buffer at all.

The pipe implementation being deprecated has a similar capability so better
have it here too.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
tests: app_kernel: restore the PIPE_NOBUFF variant

... now that the new pipe implementation supports it.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

Mattemagikern commented on kernel/pipe.c:
```diff
@@ -38,35 +38,25 @@ static int wait_for(_wait_q_t *waitq, struct k_pipe *pipe, k_spinlock_key_t *key
 	pipe->waiting++;
 	SYS_PORT_TRACING_OBJ_FUNC_BLOCKING(k_pipe, read, pipe, timeout);
-	z_pend_curr(&pipe->lock, *key, waitq, timeout);
+	rc = z_pend_curr(&pipe->lock, *key, waitq, timeout);
```
I didn't utalize this because of this discussion with @andyross: 
https://github.com/zephyrproject-rtos/zephyr/pull/83247#issuecomment-2555967498

andyross replied:
To clarify: I'm not a fan of promiscuous use of z_thread_return_value_set as it usually pretty confusing, wastes bytes in the thread struct, and tends to just recapitulate a bunch of state that the IPC mechanisms need to keep anyway (i.e. you generally need to have some sideband to tell you "why you woke up" and this is just a duplicate).   My hope in the long term is that pend() ends up just flagging a boolean result: either you were deliberately awoken (currently a return value of zero), or you bounced for some async reason (and need to return -EAGAIN to your caller).

And in fact that's all this is doing: it just does the wakeup with a result code of zero, and checks for the -EAGAIN which is the default value.  So while this is using a feature I personally dislike, it's doing it in a way that would port cleanly to the eventual utopia I imagine.

Mattemagikern commented on kernel/pipe.c:
```diff
@@ -110,7 +100,7 @@ int z_impl_k_pipe_write(struct k_pipe *pipe, const uint8_t *data, size_t len, k_
 
 		if (pipe_empty(pipe)) {
-			notify_waiter(&pipe->data);
+			z_sched_wake(&pipe->data, 0, NULL);
```
What is the difference between the `z_sched_wake` & `z_upend_*` ?

andyross replied:
Layering.  The former is synchronized and manages the timeout value and some other bits of state you need to worry about when you wake a thread up.

andyross reviewed (dismissed):
This looks clean.  Preemptive +1 which people can assume is refreshed post-merge of the parent PR.

peter-mitsis reviewed (commented):
I may have missed some subtleties (won't be the first or last time), but pending additional info there are I think a few issues with this proposed patch set.

1. Shouldn't use of z_sched_wake() necessitate a call to z_reschedule() or one of its variants? The thread that is being awakened could be of higher priority than the current thread, so I would think that we would want to switch to it as soon as we can safely do so rather than rely upon some schedule point external to the pipe operation.
2. The way I read the pipe reading code, it looks like we can wind up in a situation where we have a non-full pipe buffer and waiting writers.  Imagine a full pipe buffer (say 100 bytes) with two waiting writers (that each want to write 1 byte) and we read exactly 100 bytes. The first writer will wake, the pipe buffer will be read, eventually the awakened writer will write its 1 byte to the pipe buffer, but no one wakes the 2nd writer.

npitre committed:
kernel/pipe: add missing calls to z_reschedule()

We are waking up threads but failed to let them run if they are
higher priority. Add missing calls to z_reschedule().

Also wake up all pending writers as we don't know how many there might
be. It is more efficient to wake them all when the ring buffer is full
before reading from it rather than waking them one by one whenever there is
more room in it.

Thanks to Peter Mitsis for noticing those issues.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
both issues were real and should now be fixed.

peter-mitsis commented:
I think that we are almost there. One more thing that I noticed is that the _pipe_buf_spec_ is stored on the stack. This can be an issue on SMP platforms that do not have a coherent cache for the stack (xtensa). This is why the old pipes implementation had a copy of the pipe descriptor embedded within the thread structure (and swap_data pointed to it).  I have some more thinking to do, but there might be some potential behavioral issues or asymmetries if _k_pipe_read()_ is called from within an ISR and there are multiple writers waiting.

npitre commented:
> One more thing that I noticed is that the _pipe_buf_spec_ is stored on the
> stack. This can be an issue on SMP platforms that do not have a coherent
> cache for the stack (xtensa).

Can it be made coherent manually? The thread owning the stack is guaranteed
not to execute while other threads/CPUs access it.

This also means Xtensa cannot read from a pipe to a buffer on the stack either.

One possibility is to simply disable the direct copy path on Xtensa.

> I have some more thinking to do, but there might be some potential 
> behavioral issues or asymmetries if `k_pipe_read()` is called from within
> an ISR and there are multiple writers waiting.

`k_pipe_read()` from within an ISR cannot sleep. It may only consume data from
the ring buffer if any, and if the ring buffer was full it wakes up
potential waiting writers, and returns immediately.

andyross commented:
FWIW it's not Xtensa as a whole, it's specifically intel_adsp's SMP configurations.  The architecture actually has a tunable for snoop-enabled cache, but I'm not aware of anyone that's used it.

And yes: it's best to just gate the feature on `!defined(CONFIG_KERNEL_COHERENCE)`.  It's possible to manually manage the cache, but very non-trivial.  Not only is it incoherent, there's no bytewise tagging: so if one side writes one byte of a 64 byte cache line and ignores the other 63, a future flush can still clobber work other contexts have done.  So you have to align and pad everything.  **And** the cache line size changes between cores (the MTK ones I'm working with right now have 128 byte lines!).  Not really worth it.

Honestly that's sort of my feeling for all the zero-copy work in our existing IPC primitives.  Feels like a lot of code for not a lot of benefit.  Apps that care about performance on that level should probably be passing pointers and queuing them with semaphores.  Using a pipe almost always means you're willing to deal with some inefficiency in order to get a clean API.

npitre committed:
kernel/pipe: disable direct-to-readers copy with CONFIG_KERNEL_COHERENCE

Systems that enabled this option don't have their stacks in coherent
memory. Given our pipe_buf_spec is stored on the stack, and readers may
also have their destination buffer on their stack too, it is not worth
going to the trouble of supporting direct-to-readers copy with them.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
CONFIG_KERNEL_COHERENCE awareness added.

Mattemagikern commented:
> add back the ability to do direct writer-to-reader transfers and no-ring-buffer operations as the deprecated code did.

Do we really want to do this?
Traditionally, pipes serve as buffers between producers and consumers, ensuring a clear separation of concerns. A copy-based implementation guarantees consistent semantics regardless of producer/consumer timing or priority, making it the expected behavior of a pipe.

While the NULL buffer option gives users more flexibility, we should ask whether this optimization is truly worth it and widely applicable. Would it see enough real-world use to justify the added complexity? More importantly, are there better alternatives that should be used instead?

By not offering this option, we might actually guide users toward a more suitable data structure for zero-copy communication, rather than complicating the pipe abstraction.

I'm not necessarily opposed to this change—I just want to make sure we've fully considered the implications.

npitre commented:
> Do we really want to do this?

Absolutely.

That is, in fact, the most likely mode of operation in a client-server 
scenario where the actual ring buffer would never be touched unless the 
server cannot keep up. By not doing this, the pipe transfer overhead is 
literally doubled.

As for the added code complexity claim goes, we now have something that 
is magnitude simpler than the code this is replacing. This particular 
feature holds in a rather smallish function of its own.

If you want to steer people towards "better alternatives" you don't do 
it by crippling existing mechanisms. Sometimes it is just a matter of 
making those existing mechanisms good enough, and raising the bar for 
alternatives is certainly a good thing overall.

peter-mitsis reviewed (approved):
This seems to be surviving everything I have been throwing at it to try and break it. :)

kartben commented:
@npitre et al this seems to be breaking CI on main (SMP only, maybe). Could someone please have a look?

https://github.com/zephyrproject-rtos/zephyr/actions/runs/12893726400/job/35950760126


```
west twister -p qemu_cortex_a53/qemu_cortex_a53/smp -s arch.common.stack_unwind.arm64 --no-detailed-test-id



[111/129] Building C object zephyr/kernel/CMakeFiles/kernel.dir/pipe.c.obj
FAILED: zephyr/kernel/CMakeFiles/kernel.dir/pipe.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.17.0/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=0 -DPICOLIBC_DOUBLE_PRINTF_SCANF -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/__w/zephyr/zephyr/kernel/include -I/__w/zephyr/zephyr/arch/arm64/include -I/__w/zephyr/zephyr/twister-out/qemu_cortex_a53_qemu_cortex_a53_smp/zephyr/tests/arch/common/stack_unwind/arch.common.stack_unwind.symtab/zephyr/include/generated/zephyr -I/__w/zephyr/zephyr/include -I/__w/zephyr/zephyr/twister-out/qemu_cortex_a53_qemu_cortex_a53_smp/zephyr/tests/arch/common/stack_unwind/arch.common.stack_unwind.symtab/zephyr/include/generated -I/__w/zephyr/zephyr/soc/arm -I/__w/zephyr/zephyr/soc/arm/qemu_cortex_a53/. -I/__w/zephyr/zephyr/subsys/testsuite/include -isystem /__w/zephyr/zephyr/lib/libc/common/include -Wshadow -fno-strict-aliasing -Werror -Og -imacros /__w/zephyr/zephyr/twister-out/qemu_cortex_a53_qemu_cortex_a53_smp/zephyr/tests/arch/common/stack_unwind/arch.common.stack_unwind.symtab/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-a53 -mabi=lp64 --sysroot=/opt/toolchains/zephyr-sdk-0.17.0/aarch64-zephyr-elf/aarch64-zephyr-elf -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-omit-frame-pointer -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/tests/arch/common/stack_unwind=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -mno-omit-leaf-frame-pointer -moverride=tune=no_ldp_stp_qregs -specs=picolibc.specs -std=c99 -MD -MT zephyr/kernel/CMakeFiles/kernel.dir/pipe.c.obj -MF zephyr/kernel/CMakeFiles/kernel.dir/pipe.c.obj.d -o zephyr/kernel/CMakeFiles/kernel.dir/pipe.c.obj -c /__w/zephyr/zephyr/kernel/pipe.c
In file included from /__w/zephyr/zephyr/kernel/include/ksched.h:11,
                 from /__w/zephyr/zephyr/kernel/pipe.c:9:
In function 'arch_thread_return_value_set',
    inlined from 'z_thread_return_value_set_with_data' at /__w/zephyr/zephyr/kernel/include/kernel_internal.h:130:2,
    inlined from 'copy_to_pending_readers' at /__w/zephyr/zephyr/kernel/pipe.c:134:4:
/__w/zephyr/zephyr/kernel/include/kernel_internal.h:121:29: error: 'reader' may be used uninitialized [-Werror=maybe-uninitialized]
  121 |         thread->swap_retval = value;
      |         ~~~~~~~~~~~~~~~~~~~~^~~~~~~
/__w/zephyr/zephyr/kernel/pipe.c: In function 'copy_to_pending_readers':
/__w/zephyr/zephyr/kernel/pipe.c:92:26: note: 'reader' was declared here
   92 |         struct k_thread *reader;
      |                          ^~~~~~
cc1: all warnings being treated as errors
ninja: build stopped: subcommand failed.

INFO    - /__w/zephyr/zephyr/twister-out/qemu_cortex_a53_qemu_cortex_a53_smp/zephyr/tests/arch/common/stack_unwind/arch.common.stack_unwind.symtab/build.log
INFO    -  369/1153 qemu_cortex_a53/qemu_cortex_a53/smp arch.common.stack_unwind.arm64                     ERROR Build failure (qemu <zephyr>)
INFO    - /__w/zephyr/zephyr/twister-out/qemu_cortex_a53_qemu_cortex_a53_smp/zephyr/tests/arch/common/stack_unwind/arch.common.stack_unwind.arm64/build.log

```

npitre commented:
> @npitre et al. this seems to be breaking CI on main (SMP only, maybe). Could someone please have a look?

Fix in PR #84352.
